core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -617,7 +639,44 @@ protected void doRun() throws Exception {
                 finishAsFailed(e);
                 return;
             }
-            finishAndMoveToReplication(replicationPhase);
+
```

Comment:
This method is becoming a bit of a monster. How about simplifying it like this? 

https://gist.github.com/bleskes/84c46a548d2782e6064a

Note that I removed the non-existent nodes check - I really wonder when this can happen these days. Also, I rather take the shardrouting from the indexShard and know it's consistent with the IndexShard#relocated()